File: maze.py
def dfs(maze, st, dest):
        if (st[0] == dest[0] and st[1] == dest[1]):
            return True
        
        if (st[0] < 0 or st[0] >= len(maze) or st[1] < 0 or st[1] >= len(maze[0]) or maze[st[0]][st[1]] == 1):
            return False
        
        maze[st[0]][st[1]] = -1
        dirs = [[0,1],[0,-1],[1,0],[1,0]]
        for p in dirs: 
            r = p[0]
            c = p[1]
            nextr = st[0] + r
            nextc = st[1] + c
            changed = False
            while (0 <= nextc < len(maze[0]) and 0 <= nextr < len(maze) and 
                   maze[nextr][nextc] != -1 and maze[nextr][nextc] != 1):
                maze[nextr][nextc] = -1
                nextr += r
                nextc += c
                changed = True
            if changed:    
                if dfs(maze,[nextr-r,nextc-c],dest):
                    return True
        return False
    
        
def hasPath(maze, start, destination):
    """
    :type maze: List[List[int]]
    :type start: List[int]
    :type destination: List[int]
    :rtype: bool
    """
    return dfs(maze, start, destination)


maze = [[0,0,0,0,1,0,0],[0,0,1,0,0,0,0],[0,0,0,0,0,0,0],[0,0,0,0,0,0,1],[0,1,0,0,0,0,0],[0,0,0,1,0,0,0],[0,0,0,0,0,0,0],[0,0,1,0,0,0,1],[0,0,0,0,1,0,0]]
st = [0,0]
dest = [8,6]
print hasPath(maze,st,dest)

File: maze.cpp
#include<iostream>
#include<vector>
using namespace std;


bool dfs(vector<vector<int>>& maze, vector<int>& st, vector<int>& dest, int rows, int cols) {
    if (st[0] == dest[0] && st[1] == dest[1])
        return true;
    
    if (st[0] < 0 || st[0] >= maze.size() || st[1] < 0 || st[1] >= maze[0].size() || maze[st[0]][st[1]] == -1)
        return false;
    maze[st[0]][st[1]] = -1;
    vector<vector<int>> dirs = {{0,1},{0,-1},{1,0},{-1,0}};
    for (auto p : dirs) {
        int r = p[0];
        int c = p[1];
        while (c < maze[0].size() && maze[r][c] != -1 && maze[r][c] != 1 && r >= 0 && r < maze.size()) {
            st[0] += r;
            st[1] += c;
            maze[st[0]][st[1]] = -1;
        }
        if (dfs(maze,st,dest,rows,cols))
            return true;  
    }
    return false;
}
bool hasPath(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
    int rows = maze.size();
    int cols = maze[0].size();
    //maze[start[0]][start[1]] = -1;
    return dfs(maze, start, destination, rows,cols);
    
}


int main()
{
    vector<vector<int>> maze = {{0,0,1,0,0},{0,0,0,0,0},{0,0,0,1,0},{1,1,0,1,1},{0,0,0,0,0}};
    vector<int> start = {0,4};
    vector<int> dest = {4,4};
    cout << hasPath(maze,start,dest);

}
